### PR TITLE
[4.0] [com_contact] convert to prepared statement

### PR DESCRIPTION
**Pull request by alikon**
### Summary of Changes
use prepared statement for SQL

### Testing Instructions
test com_contact

### Expected result
should work as before

### Actual result

don't use prepared statements

### CODE REVIEW

File: administrator/components/com_contact/Model/ContactModel.php [hunk 1]
```diff
@@ -22,6 +22,8 @@
 use Joomla\Registry\Registry;
 use Joomla\String\StringHelper;
 use Joomla\Utilities\ArrayHelper;
+use Joomla\Database\ParameterType;
```
Comment (Quy) [inline administrator/components/com_contact/Model/ContactModel.php]: Sort alpha order. Repeat in other files.

File: components/com_contact/Model/CategoryModel.php [hunk 2]
```diff
@@ -147,13 +147,14 @@ protected function getListQuery()
 		 */
 			->from($db->quoteName('#__contact_details', 'a'))
 			->leftJoin($db->quoteName('#__categories', 'c') . ' ON c.id = a.catid')
-			->where('a.access IN (' . $groups . ')');
+			->whereIn($db->quoteName('a.access'), $groups );
```
Comment (Quy) [inline components/com_contact/Model/CategoryModel.php]:
```suggestion
			->whereIn($db->quoteName('a.access'), $groups);
```

File: components/com_contact/Model/CategoryModel.php [hunk 3]
```diff
@@ -147,13 +147,14 @@ protected function getListQuery()
...
 
 		// Filter by category.
 		if ($categoryId = $this->getState('category.id'))
 		{
-			$query->where('a.catid = ' . (int) $categoryId)
-				->where('c.access IN (' . $groups . ')');
+			$query->bind(':acatid', $categoryId, ParameterType::INTEGER);
+			$query->where('a.catid = :acatid')
+				->whereIn($db->quoteName('c.access'), $groups );
```
Comment (Quy) [inline components/com_contact/Model/CategoryModel.php]:
```suggestion
				->whereIn($db->quoteName('c.access'), $groups);
```

File: administrator/components/com_contact/Model/ContactModel.php (hunk 1 above)
Comment (alikon) [reply administrator/components/com_contact/Model/ContactModel.php]: done

File: administrator/components/com_contact/Model/ContactsModel.php [hunk 4]
```diff
@@ -254,22 +255,23 @@ protected function getListQuery()
 		// Filter by access level.
 		if ($access = $this->getState('filter.access'))
 		{
-			$query->where($db->quoteName('a.access') . ' = ' . (int) $access);
+			$query->bind(':access', $access);
```
Comment (Quy) [inline administrator/components/com_contact/Model/ContactsModel.php]:
```suggestion
			$query->bind(':access', $access, ParameterType::INTEGER);
```

File: components/com_contact/Model/CategoryModel.php [hunk 5]
```diff
@@ -167,21 +168,21 @@ protected function getListQuery()
...
 		// Filter by start and end dates.
-		$nullDate = $db->quote($db->getNullDate());
-		$nowDate = $db->quote(Factory::getDate()->toSql());
+		$nowDate = Factory::getDate()->toSql();
 
 		if ($this->getState('filter.publish_date'))
 		{
-			$query->where('(' . $query->isNullDatetime('a.publish_up') . ' OR a.publish_up <= ' . $db->quote($nowDate) . ')')
-				->where('(' . $query->isNullDatetime('a.publish_down') . ' OR a.publish_down >= ' . $db->quote($nowDate) . ')');
+			$query->where('(' . $query->isNullDatetime('a.publish_up') . ' OR a.publish_up <= :publish_up)')
```
Comment (Quy) [inline components/com_contact/Model/CategoryModel.php]: Add binding.

File: components/com_contact/Model/FeaturedModel.php [hunk 6]
```diff
@@ -83,7 +84,7 @@ public function getItems()
 	protected function getListQuery()
 	{
 		$user = Factory::getUser();
-		$groups = implode(',', $user->getAuthorisedViewLevels());
+		$groups =  $user->getAuthorisedViewLevels();
```
Comment (Quy) [inline components/com_contact/Model/FeaturedModel.php]:
```suggestion
		$groups = $user->getAuthorisedViewLevels();
```

File: administrator/components/com_contact/Field/Modal/ContactField.php [hunk 7]
```diff
@@ -17,6 +17,8 @@
 use Joomla\CMS\Language\LanguageHelper;
 use Joomla\CMS\Language\Text;
 use Joomla\CMS\Session\Session;
+use Joomla\Database\ParameterType;
+
```
Comment (Quy) [inline administrator/components/com_contact/Field/Modal/ContactField.php:21]:
```suggestion
```

File: administrator/components/com_contact/Model/ContactsModel.php [hunk 8]
```diff
@@ -295,29 +298,34 @@ protected function getListQuery()
...
+					'(' . $db->quoteName('a.name') . ' LIKE :name OR ' . $db->quoteName('a.alias') . ' LIKE :alias' . ')'
 				);
 			}
 		}
 
 		// Filter on the language.
 		if ($language = $this->getState('filter.language'))
 		{
-			$query->where($db->quoteName('a.language') . ' = ' . $db->quote($language));
+			$query->bind(':language', $db->quote($language));
```
Comment (Quy) [inline administrator/components/com_contact/Model/ContactsModel.php]: Remove `$db->quote`?

File: components/com_contact/Model/CategoryModel.php [hunk 9]
```diff
@@ -167,21 +168,21 @@ protected function getListQuery()
...
 		if (is_numeric($state))
 		{
-			$query->where('a.published = ' . (int) $state);
+			$query->bind(':published', $state, ParameterType::INTEGER);
+			$query->where('a.published = :published');
 		}
 		else
 		{
-			$query->where('(a.published IN (0,1,2))');
+			$query->whereIn($db->quoteName('c.published'), [0, 1, 2]);
```
Comment (Quy) [inline components/com_contact/Model/CategoryModel.php]:
```suggestion
			$query->whereIn($db->quoteName('c.published'), [0,1,2]);
```

File: components/com_contact/Model/CategoryModel.php [hunk 10]
```diff
@@ -191,12 +192,16 @@ protected function getListQuery()
...
+			$query->where($db->quoteName('a.name') . ' LIKE :name ');
 		}
 
-		// Filter by language
-		if ($this->getState('filter.language'))
+		// Filter on the language.
+		if ($language = $this->getState('filter.language'))
 		{
-			$query->where('a.language in (' . $db->quote(Factory::getLanguage()->getTag()) . ',' . $db->quote('*') . ')');
+			$query->bind(':language', $db->quote($language));
```
Comment (Quy) [inline components/com_contact/Model/CategoryModel.php]: Remove `$db->quote`?

File: components/com_contact/Model/CategoryModel.php (hunk 5 above)
Comment (alikon) [reply components/com_contact/Model/CategoryModel.php]: done thanks

File: components/com_contact/Model/ContactModel.php [hunk 11]
```diff
@@ -98,41 +99,6 @@ public function getForm($data = array(), $loadData = true)
 			return false;
 		}
 
-		$temp = clone $this->getState('params');
```
Comment (wilsonge) [inline components/com_contact/Model/ContactModel.php]: Why is this all removed? I think for the purposes of prepared statements even if it's not used i'd rather see this in another PR please

File: administrator/components/com_contact/Model/ContactsModel.php (hunk 8 above)
Comment (alikon) [reply administrator/components/com_contact/Model/ContactsModel.php]: done

File: components/com_contact/Model/CategoryModel.php (hunk 10 above)
Comment (alikon) [reply components/com_contact/Model/CategoryModel.php]: done

File: components/com_contact/Model/ContactModel.php (hunk 11 above)
Comment (alikon) [reply components/com_contact/Model/ContactModel.php]: duplicated code  https://github.com/joomla/joomla-cms/blob/staging/components/com_contact/views/contact/view.html.php#L87

File: components/com_contact/Model/CategoryModel.php [hunk 12]
```diff
@@ -191,12 +194,16 @@ protected function getListQuery()
...
+			$query->bind(':name', $search);
+			$query->where($db->quoteName('a.name') . ' LIKE :name ');
 		}
 
-		// Filter by language
-		if ($this->getState('filter.language'))
+		// Filter on the language.
+		if ($language = $this->getState('filter.language'))
 		{
-			$query->where('a.language in (' . $db->quote(Factory::getLanguage()->getTag()) . ',' . $db->quote('*') . ')');
```
Comment (SharkyKZ) [inline components/com_contact/Model/CategoryModel.php:199]: All languages (*) removed by mistake?

**Comment by SharkyKZ** [comment]
What's the consensus on throwing exceptions vs using `setError()` and returning `false` in models? I see a mix of both in same model, e.g. https://github.com/joomla/joomla-cms/blob/48e6af8c88c93df14ef6535b9db23d46fe649ba1/components/com_contact/Model/ContactModel.php#L194-L197 and https://github.com/joomla/joomla-cms/blob/48e6af8c88c93df14ef6535b9db23d46fe649ba1/components/com_contact/Model/ContactModel.php#L251-L253

**Comment by SharkyKZ** [comment]
Nevermind, exception is thrown inside try/catch.

File: administrator/components/com_contact/Field/Modal/ContactField.php [hunk 13]
```diff
@@ -108,7 +109,8 @@ function jSelectContact_" . $this->id . "(id, title, object) {
 			$query = $db->getQuery(true)
 				->select($db->quoteName('name'))
 				->from($db->quoteName('#__contact_details'))
-				->where($db->quoteName('id') . ' = ' . (int) $value);
+				->bind(':id', $value, ParameterType::INTEGER)
```
Comment (Quy) [inline administrator/components/com_contact/Field/Modal/ContactField.php]: Move bind. Repeat.

File: components/com_contact/Model/ContactModel.php (hunk 11 above)
Comment (wilsonge) [reply components/com_contact/Model/ContactModel.php]: OK thats for another PR please (also imo that code belongs in the model not the view anyway)

File: components/com_contact/Model/CategoryModel.php [hunk 14]
```diff
@@ -147,13 +147,14 @@ protected function getListQuery()
...
-			->where('a.access IN (' . $groups . ')');
+			->whereIn($db->quoteName('a.access'), $groups);
 
 		// Filter by category.
 		if ($categoryId = $this->getState('category.id'))
 		{
-			$query->where('a.catid = ' . (int) $categoryId)
-				->where('c.access IN (' . $groups . ')');
+			$query->bind(':acatid', $categoryId, ParameterType::INTEGER);
+			$query->where('a.catid = :acatid')
```
Comment (HLeithner) [inline components/com_contact/Model/CategoryModel.php]: quoteName

File: administrator/components/com_contact/Field/Modal/ContactField.php (hunk 13 above)
Comment (HLeithner) [reply administrator/components/com_contact/Field/Modal/ContactField.php]: move bind to last call in chain

File: components/com_contact/Model/CategoryModel.php [hunk 15]
```diff
@@ -167,21 +168,23 @@ protected function getListQuery()
 
 		if (is_numeric($state))
 		{
-			$query->where('a.published = ' . (int) $state);
+			$query->bind(':published', $state, ParameterType::INTEGER);
+			$query->where('a.published = :published');
```
Comment (HLeithner) [inline components/com_contact/Model/CategoryModel.php]: quoteName

File: components/com_contact/Model/CategoryModel.php [hunk 16]
```diff
@@ -167,21 +168,23 @@ protected function getListQuery()
...
 		// Filter by start and end dates.
-		$nullDate = $db->quote($db->getNullDate());
-		$nowDate = $db->quote(Factory::getDate()->toSql());
+		$nowDate = Factory::getDate()->toSql();
 
 		if ($this->getState('filter.publish_date'))
 		{
-			$query->where('(' . $query->isNullDatetime('a.publish_up') . ' OR a.publish_up <= ' . $db->quote($nowDate) . ')')
-				->where('(' . $query->isNullDatetime('a.publish_down') . ' OR a.publish_down >= ' . $db->quote($nowDate) . ')');
+			$query->bind(':publish_up', $nowDate)
```
Comment (HLeithner) [inline components/com_contact/Model/CategoryModel.php]: move bind to last call in chain

File: components/com_contact/Model/CategoryModel.php [hunk 17]
```diff
@@ -167,21 +168,23 @@ protected function getListQuery()
...
+				->bind(':publish_down', $nowDate)
```
Comment (HLeithner) [inline components/com_contact/Model/CategoryModel.php]: move bind to last call in chain

File: components/com_contact/Model/CategoryModel.php [hunk 18]
```diff
@@ -167,21 +168,23 @@ protected function getListQuery()
...
+				->where('(' . $query->isNullDatetime('a.publish_up') . ' OR a.publish_up <= :publish_up)')
```
Comment (HLeithner) [inline components/com_contact/Model/CategoryModel.php]: quoteName

File: components/com_contact/Model/CategoryModel.php [hunk 19]
```diff
@@ -167,21 +168,23 @@ protected function getListQuery()
...
+				->where('(' . $query->isNullDatetime('a.publish_down') . ' OR a.publish_down >= :publish_down)');
```
Comment (HLeithner) [inline components/com_contact/Model/CategoryModel.php]: quoteName

File: components/com_contact/Model/CategoryModel.php [hunk 20]
```diff
@@ -191,12 +194,16 @@ protected function getListQuery()
 		{
 			$search = $db->quote('%' . $db->escape($search, true) . '%');
 			$query->where('(a.name LIKE ' . $search . ')');
+			$search = '%' . $db->escape(trim($search), true) . '%';
```
Comment (HLeithner) [inline components/com_contact/Model/CategoryModel.php]: escape shouldn't be needed

File: components/com_contact/Model/CategoryModel.php [hunk 21]
```diff
@@ -191,12 +194,16 @@ protected function getListQuery()
...
+			$query->bind(':name', $search);
```
Comment (HLeithner) [inline components/com_contact/Model/CategoryModel.php]: move bind to last call in chain

File: components/com_contact/Model/CategoryModel.php [hunk 22]
```diff
@@ -191,12 +194,16 @@ protected function getListQuery()
...
+			$query->where($db->quoteName('a.name') . ' LIKE :name ');
 		}
 
-		// Filter by language
-		if ($this->getState('filter.language'))
+		// Filter on the language.
+		if ($language = $this->getState('filter.language'))
 		{
-			$query->where('a.language in (' . $db->quote(Factory::getLanguage()->getTag()) . ',' . $db->quote('*') . ')');
+			$query->bind(':language', $language);
```
Comment (HLeithner) [inline components/com_contact/Model/CategoryModel.php]: move bind to last call in chain and change to whereIn + '*' as @SharkyKZ mentioned

File: components/com_contact/Model/ContactModel.php [hunk 23]
```diff
@@ -197,32 +163,35 @@ public function getItem($pk = null)
 					->select('c.title AS category_title, c.alias AS category_alias, c.access AS category_access')
 					->leftJoin($db->quoteName('#__categories', 'c') . ' ON c.id = a.catid')
 
-
 					// Join over the categories to get parent category titles
 					->select('parent.title AS parent_title, parent.id AS parent_id, parent.path AS parent_route, parent.alias AS parent_alias')
 					->leftJoin($db->quoteName('#__categories', 'parent') . ' ON parent.id = c.parent_id')
-
-					->where('a.id = ' . (int) $pk);
+					->bind(':id', $pk, ParameterType::INTEGER)
```
Comment (HLeithner) [inline components/com_contact/Model/ContactModel.php]: move bind to last call in chain

File: components/com_contact/Model/ContactModel.php [hunk 24]
```diff
@@ -197,32 +163,35 @@ public function getItem($pk = null)
...
 				if (is_numeric($published))
 				{
-					$query->where('(a.published = ' . (int) $published . ' OR a.published =' . (int) $archived . ')')
-						->where('(' . $query->isNullDatetime('a.publish_up') . ' OR a.publish_up <= ' . $db->quote($nowDate) . ')')
-						->where('(' . $query->isNullDatetime('a.publish_down') . ' OR a.publish_down >= ' . $db->quote($nowDate) . ')');
+					$query->bind(':published', $published, ParameterType::INTEGER)
+						->bind(':archived', $archived, ParameterType::INTEGER)
+						->bind(':publish_up', $nowDate)
+						->bind(':publish_down', $nowDate)
+						->where('(a.published = :published OR a.published = :archived)')
```
Comment (HLeithner) [inline components/com_contact/Model/ContactModel.php]: quoteName

File: components/com_contact/Model/ContactModel.php [hunk 25]
```diff
@@ -197,32 +163,35 @@ public function getItem($pk = null)
...
+						->where('(' . $query->isNullDatetime('a.publish_up') . ' OR a.publish_up <= :publish_up)')
```
Comment (HLeithner) [inline components/com_contact/Model/ContactModel.php]: quoteName

File: components/com_contact/Model/ContactModel.php [hunk 26]
```diff
@@ -197,32 +163,35 @@ public function getItem($pk = null)
...
+						->where('(' . $query->isNullDatetime('a.publish_down') . ' OR a.publish_down >= :publish_down)');
```
Comment (HLeithner) [inline components/com_contact/Model/ContactModel.php]: quoteName

File: components/com_contact/Model/ContactModel.php [hunk 27]
```diff
@@ -197,32 +163,35 @@ public function getItem($pk = null)
...
 				}
 
 				$db->setQuery($query);
 				$data = $db->loadObject();
 
-				if (empty($data))
+				if ($data === null)
```
Comment (HLeithner) [inline components/com_contact/Model/ContactModel.php]: Why is this change in this PR?

File: components/com_contact/Model/FeaturedModel.php [hunk 28]
```diff
@@ -93,14 +94,15 @@ protected function getListQuery()
 		$query->select($this->getState('list.select', 'a.*'))
 			->from($db->quoteName('#__contact_details', 'a'))
 			->where('a.featured = 1')
-			->where('a.access IN (' . $groups . ')')
+			->whereIn($db->quoteName('a.access'), $groups)
 			->innerJoin($db->quoteName('#__categories', 'c') . ' ON c.id = a.catid')
```
Comment (HLeithner) [inline components/com_contact/Model/FeaturedModel.php:98]: quoteName

File: components/com_contact/Model/FeaturedModel.php [hunk 29]
```diff
@@ -93,14 +94,15 @@ protected function getListQuery()
 		$query->select($this->getState('list.select', 'a.*'))
```
Comment (HLeithner) [inline components/com_contact/Model/FeaturedModel.php:94]: rewrite so something useful maybe something with quoteName

File: components/com_contact/Model/FeaturedModel.php [hunk 30]
```diff
@@ -93,14 +94,15 @@ protected function getListQuery()
...
+			->whereIn($db->quoteName('a.access'), $groups)
 			->innerJoin($db->quoteName('#__categories', 'c') . ' ON c.id = a.catid')
-			->where('c.access IN (' . $groups . ')');
+			->whereIn($db->quoteName('c.access'), $groups);
 
 		// Filter by category.
 		if ($categoryId = $this->getState('category.id'))
 		{
-			$query->where('a.catid = ' . (int) $categoryId);
+			$query->bind(':catid', $categoryId, ParameterType::INTEGER)
```
Comment (HLeithner) [inline components/com_contact/Model/FeaturedModel.php]: move bind to last call in chain

File: components/com_contact/Model/FeaturedModel.php [hunk 31]
```diff
@@ -111,21 +113,24 @@ protected function getListQuery()
 
 		if (is_numeric($state))
 		{
-			$query->where('a.published = ' . (int) $state);
+			$query->bind(':published', $state, ParameterType::INTEGER)
```
Comment (HLeithner) [inline components/com_contact/Model/FeaturedModel.php]: move bind to last call in chain

File: components/com_contact/Model/FeaturedModel.php [hunk 32]
```diff
@@ -111,21 +113,24 @@ protected function getListQuery()
...
 
 			// Filter by start and end dates.
-			$nullDate = $db->quote($db->getNullDate());
 			$date = Factory::getDate();
-			$nowDate = $db->quote($date->toSql());
-
-			$query->where('(' . $query->isNullDatetime('a.publish_up') . ' OR a.publish_up <= ' . $db->quote($nowDate) . ')')
-				->where('(' . $query->isNullDatetime('a.publish_down') . ' OR a.publish_down >= ' . $db->quote($nowDate) . ')');
+			$nowDate = $date->toSql();
+			$query->bind(':publish_up', $nowDate)
```
Comment (HLeithner) [inline components/com_contact/Model/FeaturedModel.php]: move bind to last call in chain

File: components/com_contact/Model/FeaturedModel.php [hunk 33]
```diff
@@ -111,21 +113,24 @@ protected function getListQuery()
...
+				->bind(':publish_down', $nowDate)
```
Comment (HLeithner) [inline components/com_contact/Model/FeaturedModel.php]: move bind to last call in chain

File: components/com_contact/Model/FeaturedModel.php [hunk 34]
```diff
@@ -111,21 +113,24 @@ protected function getListQuery()
...
+				->where('(' . $query->isNullDatetime('a.publish_up') . ' OR a.publish_up <= :publish_up)')
```
Comment (HLeithner) [inline components/com_contact/Model/FeaturedModel.php]: quoteName

File: components/com_contact/Model/FeaturedModel.php [hunk 35]
```diff
@@ -111,21 +113,24 @@ protected function getListQuery()
...
+				->where('(' . $query->isNullDatetime('a.publish_down') . ' OR a.publish_down >= :publish_down)');
```
Comment (HLeithner) [inline components/com_contact/Model/FeaturedModel.php]: quoteName

File: components/com_contact/View/Contact/HtmlView.php [hunk 36]
```diff
@@ -118,6 +118,12 @@ public function display($tpl = null)
 		$this->form = $this->get('Form');
 		$params     = $state->get('params');
 
+		// Check for errors.
```
Comment (HLeithner) [inline components/com_contact/View/Contact/HtmlView.php]: Why is this change in this PR?

File: administrator/components/com_contact/Model/ContactsModel.php [hunk 37]
```diff
@@ -295,41 +298,47 @@ protected function getListQuery()
...
 			if (stripos($search, 'id:') === 0)
 			{
-				$query->where('a.id = ' . (int) substr($search, 3));
+				$query->where($db->quoteName('a.id') . ' = :id');
+				$query->bind(':id', substr($search, 3), ParameterType::INTEGER);
 			}
 			else
 			{
-				$search = $db->quote('%' . str_replace(' ', '%', $db->escape(trim($search), true) . '%'));
+				$search = '%' . $db->escape(trim($search), true) . '%';
```
Comment (Quy) [inline administrator/components/com_contact/Model/ContactsModel.php]: Remove `$db->escape`?

File: components/com_contact/Model/FeaturedModel.php [hunk 38]
```diff
@@ -111,21 +113,24 @@ protected function getListQuery()
...
 		}
 
 		// Filter by language
 		if ($this->getState('filter.language'))
 		{
-			$query->where('a.language in (' . $db->quote(Factory::getLanguage()->getTag()) . ',' . $db->quote('*') . ')');
+			$language = [$db->quote(Factory::getLanguage()->getTag()), '*'];
+			$query->bind(':language', $language);
+			$query->whereIN($db->quoteName('a.language'), $language);
```
Comment (Quy) [inline components/com_contact/Model/FeaturedModel.php]:
```suggestion
			$query->whereIn($db->quoteName('a.language'), $language);
```

File: components/com_contact/Model/ContactModel.php [hunk 39]
```diff
@@ -99,11 +100,8 @@ public function getForm($data = array(), $loadData = true)
 		}
 
 		$temp = clone $this->getState('params');
-
 		$contact = $this->_item[$this->getState('contact.id')];
-
 		$active = Factory::getApplication()->getMenu()->getActive();
-
 		if ($active)
```
Comment (Quy) [inline components/com_contact/Model/ContactModel.php:106]:
```suggestion

		if ($active)
```

File: components/com_contact/Model/CategoryModel.php [hunk 40]
```diff
@@ -167,36 +169,42 @@ protected function getListQuery()
...
+			$query->bind(':name', $search);
 		}
 
-		// Filter by language
-		if ($this->getState('filter.language'))
+		// Filter on the language.
+		if ($language = $this->getState('filter.language'))
 		{
-			$query->where('a.language in (' . $db->quote(Factory::getLanguage()->getTag()) . ',' . $db->quote('*') . ')');
+			$language = [Factory::getLanguage()->getTag(), $db->quote('*')];
```
Comment (Quy) [inline components/com_contact/Model/CategoryModel.php]:
```suggestion
			$language = [Factory::getLanguage()->getTag(), '*'];
```

File: components/com_contact/Model/ContactModel.php [hunk 41]
```diff
@@ -331,21 +332,27 @@ protected function buildContactExtendedData($contact)
 				->select($this->getSlugColumn($query, 'c.id', 'c.alias') . ' AS catslug')
 				->from($db->quoteName('#__content', 'a'))
 				->leftJoin($db->quoteName('#__categories', 'c') . ' ON a.catid = c.id')
-				->where('a.created_by = ' . (int) $contact->user_id)
-				->where('a.access IN (' . $groups . ')')
+				->where($db->quoteName('a.created_by') . ' = :created_by')
+				->whereIn($db->quoteName('a.access'), $groups)
+				->bind(':created_by',  $contact->user_id, ParameterType::INTEGER)
```
Comment (Quy) [inline components/com_contact/Model/ContactModel.php]:
```suggestion
				->bind(':created_by', $contact->user_id, ParameterType::INTEGER)
```

File: administrator/components/com_contact/Model/ContactsModel.php [hunk 42]
```diff
@@ -295,41 +298,47 @@ protected function getListQuery()
...
 		}
 
 		// Filter by a single tag.
 		$tagId = $this->getState('filter.tag');
 
 		if (is_numeric($tagId))
 		{
-			$query->where($db->quoteName('tagmap.tag_id') . ' = ' . (int) $tagId)
+			$query->where($db->quoteName('tagmap.tag_id') . ' = :tag_id')
 				->join(
```
Comment (Quy) [inline administrator/components/com_contact/Model/ContactsModel.php]: Convert to `leftJoin`

File: administrator/components/com_contact/Field/Modal/ContactField.php [hunk 43]
```diff
@@ -108,7 +109,8 @@ function jSelectContact_" . $this->id . "(id, title, object) {
 			$query = $db->getQuery(true)
 				->select($db->quoteName('name'))
 				->from($db->quoteName('#__contact_details'))
-				->where($db->quoteName('id') . ' = ' . (int) $value);
+				->where($db->quoteName('id') . ' = :id')
+				->bind(':id', $value, ParameterType::INTEGER);
```
Comment (Quy) [inline administrator/components/com_contact/Field/Modal/ContactField.php:113]: Is it necessary to cast `(int)`?

Comment (alikon) [reply administrator/components/com_contact/Field/Modal/ContactField.php:113]: https://github.com/joomla/joomla-cms/blob/ee335801d91d18955a5631b7f77300960507f56e/administrator/components/com_contact/Field/Modal/ContactField.php#L106

no should run only when 1

Comment (Quy) [reply administrator/components/com_contact/Field/Modal/ContactField.php]: I mean for the bind statement.

File: administrator/components/com_contact/Model/ContactsModel.php [hunk 44]
```diff
@@ -295,41 +298,45 @@ protected function getListQuery()
...
 		$tagId = $this->getState('filter.tag');
 
 		if (is_numeric($tagId))
 		{
-			$query->where($db->quoteName('tagmap.tag_id') . ' = ' . (int) $tagId)
-				->join(
-					'LEFT',
-					$db->quoteName('#__contentitem_tag_map', 'tagmap')
+			$query->where($db->quoteName('tagmap.tag_id') . ' = :tag_id')
+				->leftjoin($db->quoteName('#__contentitem_tag_map', 'tagmap')
```
Comment (Quy) [inline administrator/components/com_contact/Model/ContactsModel.php]:
```suggestion
				->leftJoin($db->quoteName('#__contentitem_tag_map', 'tagmap')
```

File: administrator/components/com_contact/Field/Modal/ContactField.php (hunk 43 above)
Comment (Quy) [reply administrator/components/com_contact/Field/Modal/ContactField.php:113]: See https://github.com/joomla/joomla-cms/pull/25057#discussion_r289645938

Comment (HLeithner) [reply administrator/components/com_contact/Field/Modal/ContactField.php:113]: Value get already casted in line 58

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**